src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java

```diff
@@ -174,47 +186,46 @@ public void afterCommand() throws AbruptExitException {
   }
 
   /**
-   * Decide whether to keep this extension or not depending on all of:
+   * Keep an extension if and only if:
    *
    * <ol>
-   *   <li>If its dependency on os & arch didn't change
-   *   <li>If its usages haven't changed
+   *   <li>it still has usages
+   *   <li>its dependency on os & arch didn't change
+   *   <li>it hasn't become reproducible
    * </ol>
    *
+   * We do not check for changes in the usage hash of the extension or e.g. hashes of files accessed
+   * during the evaluation. These values are checked in SingleExtensionEvalFunction.
+   *
    * @param lockedExtensionKey object holding the old extension id and state of os and arch
-   * @param oldUsagesDigest the digest of usages of this extension in the existing lockfile
+   * @param newExtensionInfo the in-memory lockfile entry produced by the most recent up-to-date
+   *     evaluation of this extension (if any)
+   * @param depGraphValue the dep graph value
    * @return True if this extension should still be in lockfile, false otherwise
    */
   private boolean shouldKeepExtension(
-      ModuleExtensionId extensionId,
+      ModuleExtensionId moduleExtensionId,
       ModuleExtensionEvalFactors lockedExtensionKey,
-      byte[] oldUsagesDigest,
-      @Nullable LockFileModuleExtension.WithFactors newExtensionInfo) {
-
-    // If there is a new event for this extension, compare it with the existing ones
-    if (newExtensionInfo != null) {
-      boolean doNotLockExtension = !newExtensionInfo.moduleExtension().shouldLockExtension();
-      boolean dependencyOnOsChanged =
-          lockedExtensionKey.getOs().isEmpty()
-              != newExtensionInfo.extensionFactors().getOs().isEmpty();
-      boolean dependencyOnArchChanged =
-          lockedExtensionKey.getArch().isEmpty()
-              != newExtensionInfo.extensionFactors().getArch().isEmpty();
-      if (doNotLockExtension || dependencyOnOsChanged || dependencyOnArchChanged) {
-        return false;
-      }
+      @Nullable LockFileModuleExtension.WithFactors newExtensionInfo,
+      BazelDepGraphValue depGraphValue) {
+    if (!depGraphValue.getExtensionUsagesTable().containsRow(moduleExtensionId)) {
```

Comment:
couldn't you just read this out of `moduleResolutionEvent` as before (old line 216)? unless we're looking to stop using that event and switch to using `depGraphValue` anyway

Reply:
Yes, that's going to happen in the follow-up PR anyway (which will get even more values from Skyframe).